### PR TITLE
feat: Support field-ref on lists of structs

### PR DESCRIPTION
**Pull request by bjchambers**
A similar strategy should work for maps, but I didn't implement it yet.

### CODE REVIEW

File: crates/sparrow-compiler/src/ast_to_dfg.rs [hunk 1]
```diff
@@ -187,40 +189,14 @@ pub fn add_to_dfg(
...
-                    // to `json_field` instructions that take a `string` and output a `string`,
-                    // hence the `utf8` return type here.
-                    &DataType::Utf8
-                }
-                FenlType::Error => {
-                    // The original error is already reported.
+            let field_type = match field_type(field, base_type.inner(), arguments[0].location()) {
+                Ok(Some(field_type)) => field_type,
+                Ok(None) => {
+                    // alreday reported error.
```
Comment (bjchambers) [inline crates/sparrow-compiler/src/ast_to_dfg.rs:195]: typo

File: crates/sparrow-compiler/src/ast_to_dfg.rs [hunk 2]
```diff
@@ -896,3 +872,48 @@ fn missing_field_diagnostic(
...
+            // to `json_field` instructions that take a `string` and output a `string`,
+            // hence the `utf8` return type here.
+            Ok(Some(DataType::Utf8))
+        }
+        FenlType::Error => {
+            // The original error is already reported.
+            Ok(None)
+        }
+        _ => {
+            if let Some(args) = base_type.collection_args(&Collection::List) {
```
Comment (bjchambers) [inline crates/sparrow-compiler/src/ast_to_dfg.rs:905]: Map would need to be added here

File: crates/sparrow-instructions/src/evaluators/field_ref.rs [hunk 3]
```diff
@@ -13,6 +17,36 @@ use crate::{Evaluator, EvaluatorFactory, StaticInfo};
 pub struct FieldRefEvaluator {
     base: ValueRef,
     field_index: usize,
+    field: FieldRef,
+}
+
+fn get_field_index(field_name: &str, base: &DataType) -> anyhow::Result<(usize, FieldRef)> {
+    let (base_fields, name) = if let DataType::Struct(fields) = base {
+        (fields, None)
+    } else if let DataType::List(field) = base {
```
Comment (bjchambers) [inline crates/sparrow-instructions/src/evaluators/field_ref.rs:26]: Map would need to be added here

File: crates/sparrow-compiler/src/types/instruction.rs [hunk 4]
```diff
@@ -139,3 +114,32 @@ pub(crate) fn typecheck_inst(
...
+                }
+            })
+            .with_context(|| {
+                format!("No field named '{}' found in struct {:?}", field_name, base)
+            })?;
+
+        Ok(result_type)
+    } else if let Some(args) = base.collection_args(&sparrow_syntax::Collection::List) {
+        let field = typecheck_field_ref(field_name, &args[0])?;
+        Ok(FenlType::Collection(Collection::List, vec![field]).normalize())
```
Comment (bjchambers) [inline crates/sparrow-compiler/src/types/instruction.rs:137]: Map would need to be added here

File: crates/sparrow-instructions/src/evaluators/field_ref.rs [hunk 5]
```diff
@@ -30,38 +64,51 @@ impl EvaluatorFactory for FieldRefEvaluator {
...
+                let field = input.column(self.field_index);
+
+                // Arrow field ref ignores the null-ness of the outer struct.
+                // To handle this, we null the field out if the struct was null.
+                let nulls = NullBuffer::union(input.nulls(), field.nulls());
+
+                let data = field.to_data().into_builder().nulls(nulls).build()?;
+                Ok(make_array(data))
+            }
+            DataType::List(_) => {
```
Comment (bjchambers) [inline crates/sparrow-instructions/src/evaluators/field_ref.rs:93]: The logic for struct and list are similar -- the only difference is creating the new list array using the same offsets. We would do the same for Map (using the same key).

**Review by jordanrfrazier — APPROVED** [review]

**Pull request merged**